@here/olp-sdk-core/lib/utils/MultiPartUploadWrapper.ts

```diff
@@ -177,9 +221,211 @@ export class MultiPartUploadWrapper {
+            const promises = [];
+
+            while (promises.length < parallelRequests) {
+                const buffer = await blobData.readBytes(offset, chunkSize);
+
+                if (buffer.byteLength === 0) {
+                    break;
+                }
+
+                chunkNumber++;
+                offset += buffer.byteLength;
+                unreadBytes -= buffer.byteLength;
+
+                promises.push(
+                    request
+                        .uploadPart({
+                            data: buffer,
+                            layerId: opts.layerId,
+                            multipartToken:
+                                startMultipartResponse.multipartToken,
+                            billingTag: opts.billingTag,
+                            contentLength: buffer.byteLength,
+                            contentType: opts.contentType,
+                            partNumber: chunkNumber,
+                            url: startMultipartResponse.uploadPartUrl
+                        })
+                        .then(uploadPartResponse => {
+                            parts.push({
+                                id: uploadPartResponse.id,
+                                number: chunkNumber
+                            });
+
+                            uploadedChunks++;
+
+                            opts.onStatus &&
+                                opts.onStatus({
+                                    chunkId: uploadPartResponse.id,
+                                    chunkNumber,
+                                    chunkSize: buffer.byteLength,
+                                    totalChunks,
+                                    uploadedChunks
+                                });
+                        })
+                );
+            }
+
+            await Promise.all(promises);
+        }
+
+        await request.completeMultipartUpload({
+            parts,
+            url: startMultipartResponse.completeUrl,
+            multipartToken: startMultipartResponse.multipartToken,
+            layerId: opts.layerId,
+            billingTag: opts.billingTag
+        });
+
+        blobData.finally && blobData.finally();
+
         return STATUS_CODES.NO_CONTENT;
     }
+
+    private isStartMultipartResponseValid(obj: {
+        multipartToken?: string;
+        uploadPartUrl?: string;
+        completeUrl?: string;
+        statusUrl?: string;
+    }): boolean {
+        return (
+            obj.multipartToken !== undefined ||
+            (obj.uploadPartUrl !== undefined && obj.completeUrl !== undefined)
+        );
+    }
+
+    private isBlobData(input: any): boolean {
+        return (
+            typeof input.size === "function" &&
+            typeof input.readBytes === "function"
+        );
+    }
+
+    private async prepareBlobData(
+        input: string | File | Blob | ArrayBufferLike | BlobData
+    ): Promise<BlobData> {
+        if (this.isBlobData(input)) {
```

Comment:
Why not `input instance of BlobData` ?

Reply:
BlobData will be only if the user will set an instance to the `upload` method. Also, the user can set a path to the file (Node.js), of an instance of `Blob` or `File` (in the browser) and in that case, we'll build BlobData by ourselves.

Reply:
`instance of BlobData ` doesn't work correctly with an abstract class and optional method `finally`.

Reply:
Interesting. Ok, then, let's leave it as is.